QA.md

```diff
@@ -0,0 +1,210 @@
+# Quality Assurance
+
+This document provides an overview of our quality assurance processes for
+Tendermint. It maps out the various concerns we have regarding software quality
+against the approaches we use to ensure quality. It also attempts to capture the
+current state of our QA process and where there are still deficiencies to be
+addressed.
+
+There is a [special QA log section](#qa-log) at the end of this document where
+we can capture references to one-off QA efforts.
+
+## Approaches
+
+We make use of the following approaches to ensure delivery of quality software,
+with associated relative costs where applicable:
+
+| Approach                  | Setup Cost   | Per-Execution Cost |
+|---------------------------|--------------|--------------------|
+| Code review               |              | Moderate/High      |
+| Static analysis           | Low          | Low                |
+| Unit testing              | Moderate     | Low                |
+| Fuzz testing              | Low/Moderate | Low                |
+| Model-based testing (MBT) | High         | Low                |
+| End-to-end (E2E) testing  | High         | Low                |
+| Testnet testing           | High         | Moderate/High      |
+| Protocol audits           |              | High               |
+
+## Concerns
+
+Tendermint QA is split across the following concerns:
+- [Correct software engineering](#correct-software-engineering)
+- Protocol correctness
+  - [Consensus](#consensus)
+  - [Mempool](#mempool)
+  - [Evidence](#evidence)
+  - [Peer exchange (PEX)](#peer-exchange-pex)
+  - [Block sync](#block-sync)
+  - [State sync](#state-sync)
+  - [Light client](#light-client)
+- [Application integration (ABCI)](#application-integration-abci)
+- [Network stability](#network-stability)
+- Operational dynamics
+  - [Crash/recovery](#crash-recovery)
+  - [Upgrading](#upgrading)
+
+### Correct software engineering
+
+| Concern                         | Code Review | Static Analysis | Unit Testing | Fuzz Testing | E2E Testing | Testnet Testing |
+|---------------------------------|-------------|-----------------|--------------|--------------|-------------|-----------------|
+| Security                        |             |                 |              |              |             |                 |
+| Memory leaks                    |             |                 |              |              |             |                 |
+| Storage leaks                   |             |                 |              |              |             |                 |
+| Error handling                  |             |                 |              |              |             |                 |
+| Serialization / deserialization |             |                 |              |              |             |                 |
```

Comment:
I don't know that this view of the problem leads to an effective prioritization of methodology, or gives us particular confidence in any of these dimensions. 

While I think the best approach to this would be to start from stability/operational/performance requirements and work backwards from that, I think it'd be useful to frame the components that are the highest priority (e.g. memory leaks in testnet/e2e testing, error handling in static analysis, etc.) rather than divide our time attempting to cover everything at every level.

Reply:
The point is not to cover every concern with every method but to eventually figure out what is the most favorable methods for what concern.

In the beginning, let's just record what methods we are using and what concerns we believe we can address with them.

Eventually, we should also identify bottlenecks. E.g., if the only method we are using to find memory leaks are testnets (which seem super expensive), it might make sense to figure out whether a different (cheaper) methods will give us similar confidence. Such an overview allows us to identify such bottlenecks.

Reply:
> While I think the best approach to this would be to start from stability/operational/performance requirements and work backwards from that

This is implicit in the correctness evaluation for all of the specific protocols that Tendermint implements.